packages/vhd-lib/src/parseVhdToBlocks.js

```diff
@@ -0,0 +1,64 @@
+import { readChunk } from '@vates/read-chunk'
+import assert from 'assert'
+import { FOOTER_SIZE, HEADER_SIZE, BLOCK_UNUSED, SECTOR_SIZE } from './_constants'
+import { fuFooter, fuHeader } from './_structs'
+import checkFooter from './checkFooter'
+import checkHeader from './_checkHeader'
+
+export default async function parseVhdToBlocks(inputStream) {
+  let streamPosition = 0
+
+  async function readStream(length) {
+    const chunk = await readChunk(inputStream, length)
+    assert.strictEqual(chunk.length, length)
+    streamPosition += chunk.length
+    return chunk
+  }
+
+  const footerBuffer = await readStream(FOOTER_SIZE)
+  const footer = fuFooter.unpack(footerBuffer)
+  checkFooter(footer)
+
+  const header = fuHeader.unpack(await readStream(HEADER_SIZE))
+  checkHeader(header, footer)
+  assert.strictEqual(header.tableOffset <= 100 * 1024 * 1024, true, 'The BAT is more that 100MB in the file')
+  await readStream(header.tableOffset - streamPosition)
+  const blockSizeBytes = header.blockSize
+  const bat = await readStream(header.maxTableEntries * 4)
+  const offsetToLBA = []
+  for (let i = 0; i < bat.byteLength; i += 4) {
+    const offsetSectors = bat.readUInt32BE(i)
+    if (offsetSectors !== BLOCK_UNUSED) {
+      const lba = blockSizeBytes * (i / 4)
+      offsetToLBA.push([offsetSectors, lba])
+    }
+  }
+  offsetToLBA.sort((a, b) => a[0] - b[0])
+  const earliestBlockInFile = offsetToLBA[0][0]*SECTOR_SIZE
+  assert.strictEqual(streamPosition <= earliestBlockInFile, true, `The BAT (${streamPosition}) is after the first block of data (${earliestBlockInFile})`)
+
+  const netBitmapSizeBytes = Math.ceil(blockSizeBytes / SECTOR_SIZE / 8)
+  const paddedBitmapSizeBytes = Math.ceil(netBitmapSizeBytes / SECTOR_SIZE) * SECTOR_SIZE
+  const expectedBitmap = Buffer.alloc(netBitmapSizeBytes, 255)
+
+  async function * blockGenerator() {
+    for (const i of offsetToLBA) {
+      const offset = i[0] * SECTOR_SIZE
+      const lba = i[1]
+      const skipBytes = offset - streamPosition
+      assert.strictEqual(skipBytes >= 0, true, 'Tried to move backwards in the input stream')
+      await readStream(skipBytes)
+      const bitmap = await readStream(paddedBitmapSizeBytes)
+      assert.strictEqual(Buffer.compare(bitmap.slice(0, netBitmapSizeBytes), expectedBitmap), 0, 'bitmap was not full of ones at LBA ' + lba)
```

Comment:
Why do you need this check? :thinking:

Reply:
There are 16 fragments of VMDK for 1 VHD block, I didn't plan to potentially eliminate some VHDK fragments if some VHD blocks were ony partially filled. I just assumed all the bitmaps would all be full of ones, and I check this assumption.

Reply:
In a (differencing) VHD file, the bitmap indicates for each sectors of a block whether it has been modified since the parent VHD.

ATM, it appears that XCP-ng/XenServer don't handle this and always set the bitmap to `1`.

Anyway, even if the bitmap contains a `0` for this sector, it contains the complete data, so I think it would be best to simply ignore it.

Reply:
Ok. I'm about to jump in a train then an airplane, I don't have time to validate a change right now.

Reply:
No problem, it was just a question :slightly_smiling_face: